lib/galaxy/tool_util/linters/inputs.py

```diff
@@ -44,6 +44,11 @@ def lint_inputs(tool_xml, lint_ctx):
             if dynamic_options is None and len(select_options) == 0:
                 lint_ctx.warn(f"No options defined for select [{param_name}]")
 
+            if param_attrib.get("display") == "checkboxes":
+                if not string_as_bool(param_attrib.get("multiple", "false")):
+                    lint_ctx.error(f'Select [{param_name}] display="checkboxes" is incompatible with multiple="false"')
+                if not string_as_bool(param_attrib.get("optional", "false")):
+                    lint_ctx.error(f'Select [{param_name}] display="checkboxes" is incompatible with optional="false"')
```

Comment:
This is not the case, we have some tools with `optional="false" display="checkboxes"` and they work as expected, e.g. https://github.com/galaxyproject/tools-iuc/blob/master/tools/breseq/breseq.xml#L145
```suggestion
```

Reply:
Yep, they "work" .. but not in a user friendly way, since the user can deselect all options and gets an error message that is not helpful at all. Same is also true for `multiple="false"` .. works but not user friendly.

So I think we should actually forbid this, or change the UI

- either by implementing a helpful error message
- or by changing the behavior of UI (make it impossible to deselect all options)

Reply:
I think these are separate issues.

> Yep, they "work" .. but not in a user friendly way, since the user can deselect all options and gets an error message that is not helpful at all.

The right solution here is to improve the error message.

> Same is also true for multiple="false" .. works but not user friendly.

I think that's different, the user expectation for a set of checkboxes is that multiple values can be selected, see e.g.: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox
So I think it's good that the linter warns for `multiple="false" display="checkboxes"` .

Reply:
> The right solution here is to improve the error message.

Not sure about this. My intuition is that we should prefer the (at least functionally) equally good solution (drop-down menu select list) that avoids that the user runs into an error at all.

Reply:
> Not sure about this. My intuition is that we should prefer the (at least functionally) equally good solution (drop-down menu select list) that avoids that the user runs into an error at all.

OK, but in that case, I think we should:

1) Expand the linter error to suggest the solution, e.g.:

```suggestion
                if not string_as_bool(param_attrib.get("optional", "false")):
                    lint_ctx.error(f'Select [{param_name}] `display="checkboxes"` is incompatible with `optional="false"`, remove the `display` attribute')
```

2) Improve the "Please verify this parameter." error message anyway, since this is used also with the drop-down menu select list.

3) Maybe deprecate the `display` attribute using the tool `profile` and let Galaxy decide which one to use:

   multiple T, optional T => checkboxes
   multiple F, optional F => radio
   otherwise => default

Reply:
Sounds like a great plan. 

Do you have a pointer where I could start looking for a fix for point 2?

Regarding step 3. Was also thinking in this direction but it appeared quite radical to me :) .. Making this completely automatically and deprecating `display` is probably not a good idea if there are many options, or? 

We could still: 

- make your suggestion the default (TT => checkboxes, FF => radio, otherwise select)
- and to allow the tool developer to overwrite it 

Question is if this leads to a more consistent UI which might also be a good aim.

Reply:
> Do you have a pointer where I could start looking for a fix for point 2?

I've added a commit here to fix it.

> Question is if this leads to a more consistent UI which might also be a good aim.

Yes, we did this some years ago with the `size` attribute for `<param type="text" />` . The UI is better placed to decide how to show the option, depending also on the screen type.

Ping @wg-uiux for opinions.